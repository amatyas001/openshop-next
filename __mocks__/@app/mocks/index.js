/**
 * Generates mock products with unique attributes.
 *
 * @param {*} count
 * @returns List of mock items with dynamic values
 */
export const mockProduct = (count = 1) => {
  let index = 0;
  let items = [];
  do {
    items.push({
      id: `mock_id_${index}`,
      amount: 1,
      color: `mock_color_${index}`,
      description: `mock_description_${index}`,
      gender: `mock_gender_${index}`,
      img: `mock_img_${index}`,
      name: `mock_name_${index}`,
      price: 1,
      review: `mock_review_${index}`,
      starrating: 1,
    });
    index++;
  } while (--count !== 0);
  return items;
};

/**
 * Generates mock products with buying amount details.
 *
 * @param {*} count Number of products to generate
 * @param {*} amount Mock amount attached to the listed products
 * @returns List of mock products with mock buying amount details
 */
export const mockProductWithBuyAmount = (count, amount = 1) => {
  return mockProduct(count).map((c) => ({
    ...c,
    buy: { amount },
  }));
};

/**
 * Generates mock products with unique price attributes.
 *
 * @param {*} count Number of products to generate
 * @param {*} value Base value of prices attached to generated products
 * @returns List of mock products with dynamic values and prices
 */
export const mockProductWithPrice = (count, value) => {
  const generatePrices = function* (base) {
    let index = 1;
    while (true) yield base * index++;
  };
  const generator = generatePrices(value);
  return mockProduct(count).map((c) => ({
    ...c,
    price: generator.next().value,
  }));
};

/**
 * Generates mock products with unique color attributes.
 * Each odd products has an array format color property.
 *
 * @param {number} count Number of products to generate
 * @returns Object with exact shape of
 * - `colors`: List of generated unique colors.
 * - `products`: List of mock items with dynamic values and colors.
 */
export const mockProductWithColor = (count) => {
  const generateColors = function* () {
    let index = 0;
    while (true) {
      yield index % 2 ? `mock_color_${index}` : [`mock_color_${index}`, `mock_color_${index + 1}`];
      index++;
    }
  };
  const generator = generateColors();
  const products = mockProduct(count).map((c, i) => ({
    ...c,
    color: generator.next().value,
  }));
  return {
    colors: [...new Set(products.reduce((a, c) => a.concat(c.color), ['All']))],
    products,
  };
};
