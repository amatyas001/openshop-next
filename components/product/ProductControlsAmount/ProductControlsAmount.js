import PropTypes from 'prop-types';
import {
  NumberInput,
  NumberInputField,
  NumberInputStepper,
  NumberIncrementStepper,
  NumberDecrementStepper,
  Flex,
} from '@chakra-ui/core';
import * as COLORS from '@app/config/colors';
import { Button } from '@app/components';
import { FaMinusSquare, FaPlusSquare } from 'react-icons/fa';

const MemoProductControlsAmount = (props) => {
  const { setDetails, details, product } = props;
  return (
    <NumberInput
      data-testid='product-amount-wrapper'
      defaultValue={0}
      focusInputOnChange={false}
      max={product.amount}
      min={0}
      onChange={(e) => setDetails(e)}
      value={details.buy.amount}
    >
      <Flex width='100%'>
        <NumberInputStepper
          borderColor='transparent'
          color={COLORS.HEADING.light}
          flexDirection='row'
          alignItems='center'
          width='100%'
        >
          <NumberIncrementStepper
            as={Button}
            data-testid='product-amount-increment'
            fontSize='2.5rem'
            height='50px'
            borderTopRightRadius='0px !important'
            borderRight='0'
            bg={COLORS.BUTTON.transparent.base}
            _hover={{
              bg: 'transparent',
              color: COLORS.BUTTON.primary.hover.bg,
            }}
            _focus={{
              bg: 'transparent',
              color: COLORS.BUTTON.primary.hover.bg,
            }}
            _selected={{
              bg: 'transparent',
              color: COLORS.BUTTON.primary.hover.bg,
            }}
            _active={{
              bg: 'transparent',
              color: COLORS.BUTTON.primary.hover.bg,
            }}
            color={COLORS.BUTTON.primary.base}
          >
            <FaPlusSquare />
          </NumberIncrementStepper>
          <NumberDecrementStepper
            as={Button}
            data-testid='product-amount-decrement'
            fontSize='2.5rem'
            height='50px'
            mt='0px !important'
            bg={COLORS.BUTTON.transparent.base}
            color={COLORS.BUTTON.primary.base}
            _hover={{
              bg: 'transparent',
              color: COLORS.BUTTON.primary.hover.bg,
            }}
            _focus={{
              bg: 'transparent',
              color: COLORS.BUTTON.primary.hover.bg,
            }}
            _selected={{
              bg: 'transparent',
              color: COLORS.BUTTON.primary.hover.bg,
            }}
            _active={{
              bg: 'transparent',
              color: COLORS.BUTTON.primary.hover.bg,
            }}
            borderRadius='0'
            borderLeft='0'
          >
            <FaMinusSquare />
          </NumberDecrementStepper>
        </NumberInputStepper>
        <NumberInputField
          backgroundColor={COLORS.BG.dark}
          border='0'
          color={COLORS.HEADING.light}
          data-testid='product-amount-input'
          fontFamily='Khand'
          fontSize='1.8rem'
          fontWeight='bold'
          isReadOnly
          mb='-4px'
          textAlign='center'
          width='100%'
          _hover={{}}
          _focus={{}}
          _active={{}}
          _selected={{}}
        />
      </Flex>
    </NumberInput>
  );
};

MemoProductControlsAmount.propTypes = {
  /**
   * Unique product object
   */
  product: PropTypes.shape({
    id: PropTypes.string.isRequired,
    amount: PropTypes.number.isRequired,
    color: PropTypes.oneOfType([PropTypes.string, PropTypes.arrayOf(PropTypes.string)]),
    description: PropTypes.string,
    gender: PropTypes.string,
    img: PropTypes.string,
    name: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
    review: PropTypes.string,
    sizes: PropTypes.arrayOf(PropTypes.string),
    starrating: PropTypes.number.isRequired,
  }).isRequired,

  /**
   * Unique product object with buying details
   */
  details: PropTypes.shape({
    id: PropTypes.string.isRequired,
    amount: PropTypes.number.isRequired,
    color: PropTypes.oneOfType([PropTypes.string, PropTypes.arrayOf(PropTypes.string)]),
    description: PropTypes.string,
    gender: PropTypes.string,
    img: PropTypes.string,
    name: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
    review: PropTypes.string,
    sizes: PropTypes.arrayOf(PropTypes.string),
    starrating: PropTypes.number.isRequired,
    buy: PropTypes.shape({
      amount: PropTypes.number.isRequired,
      size: PropTypes.string,
      color: PropTypes.string,
    }),
  }).isRequired,

  /**
   * Handler to set product buying details in wrapper
   */
  setDetails: PropTypes.func.isRequired,
};

/**
 * Renders an input field to select the amount of a specific product.
 *
 * @see https://amatyas001.github.io/openshop-next/#productcard
 */
export const ProductControlsAmount = React.memo(
  MemoProductControlsAmount,
  /* istanbul ignore next */
  (prev, next) => prev.details.buy.amount === next.details.buy.amount
);
