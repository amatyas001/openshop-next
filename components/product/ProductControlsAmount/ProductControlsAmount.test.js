import { create, act } from 'react-test-renderer';
import { mockProduct } from '@app/mocks';
import { ProductControlsAmount } from './ProductControlsAmount';

describe('<ProductControlsAmount />', () => {
  let tree, wrapper;

  const product = mockProduct(1)[0];

  const details = {
    ...product,
    buy: {
      amount: 0,
    },
  };

  const setter = jest.fn();

  beforeAll(() => {
    act(() => {
      tree = create(
        <ProductControlsAmount
          product={product}
          details={details}
          setDetails={setter}
        />
      );
    });
    wrapper = tree.root.findByProps({
      'data-testid': 'product-amount-wrapper',
    });
  });

  afterEach(() => {
    setter.mockClear();
  });

  it('should render with required props', () => {
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it("should have default value of '0'", () => {
    expect(wrapper.props.value).toEqual(0);
  });

  it("should have min value of '0'", () => {
    expect(wrapper.props.min).toEqual(0);
  });

  it('should have max value of product amount', () => {
    expect(wrapper.props.max).toEqual(product.amount);
  });

  it('should call setter function', () => {
    act(() => {
      wrapper.props.onChange(1);
    });
    expect(setter).toHaveBeenCalledTimes(1);
    expect(setter).toHaveBeenCalledWith(1);
  });
});
