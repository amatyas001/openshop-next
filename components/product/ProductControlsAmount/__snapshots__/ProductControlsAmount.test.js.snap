// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductControlsAmount /> should render with required props 1`] = `
<div
  className="css-1dm3qsp"
  data-testid="product-amount-wrapper"
>
  <div
    className="css-hboir5"
  >
    <div
      aria-hidden={true}
      className="css-1yqqx3k"
    >
      <button
        className="css-fdetwv"
        data-testid="product-amount-increment"
        href="#void"
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        role="button"
        tabIndex="0"
        type="button"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-32 252c0 6.6-5.4 12-12 12h-92v92c0 6.6-5.4 12-12 12h-56c-6.6 0-12-5.4-12-12v-92H92c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h92v-92c0-6.6 5.4-12 12-12h56c6.6 0 12 5.4 12 12v92h92c6.6 0 12 5.4 12 12v56z"
          />
        </svg>
      </button>
      <button
        className="css-1fd1i4r"
        data-testid="product-amount-decrement"
        href="#void"
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        role="button"
        tabIndex="0"
        type="button"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zM92 296c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h264c6.6 0 12 5.4 12 12v56c0 6.6-5.4 12-12 12H92z"
          />
        </svg>
      </button>
    </div>
    <input
      aria-invalid={false}
      aria-readonly={true}
      aria-valuemax={1}
      aria-valuemin={0}
      aria-valuenow={0}
      autoComplete="off"
      className="css-s3c5ne"
      data-testid="product-amount-input"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      readOnly={true}
      role="spinbutton"
      type="text"
      value={0}
    />
  </div>
</div>
`;
